### PR TITLE
feat(obj): add object draw cache support.

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

Buffer the image after drawing the object to improve the rendering performance of complex objects.

```c
void obj_draw_cache_test(void)
{
    lv_obj_t* obj = lv_obj_create(lv_scr_act());
    lv_obj_draw_cache_set_enable(obj, true);
    lv_obj_draw_cache_invalidate(obj);
    lv_obj_center(obj);
}
```

### Checkpoints
- [x] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Run `code-format.py` from the `scripts` folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the documentation

### CODE REVIEW

**Comment by FASTSHIFT** [comment]
Do I need to encapsulate a dedicated memory alloc interface?

File: src/core/lv_obj_draw_cache.c [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    # error "Need to enable 'LV_USE_SNAPSHOT'"
+#endif
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+typedef struct _lv_obj_draw_cache_t {
+    lv_img_dsc_t img_dsc;
+    bool invalidate;
```
Comment (kisvegabor) [inline src/core/lv_obj_draw_cache.c]:
```suggestion
    bool invalid;
```

File: src/core/lv_obj_draw_cache.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    bool draw_normal;
```
Comment (kisvegabor) [inline src/core/lv_obj_draw_cache.c]:
```suggestion
    bool skip_cache;
```

**Comment by kisvegabor** [comment]
Nice and clean implementation. Looks great! 

I had two suggestions about naming but nothing serious. 

I've already prepared a patch with some refactoring. I'll create in minutes. I'd like to merge that first as it affects to whole library and then we can easily refactor this patch if needed.

**Comment by kisvegabor** [comment]
PR is opened: #3501

**Comment by kisvegabor** [comment]
Although Kconfig and Micropython is not updated yet I merged #3501 to not block features like this. Could you solve the minor conflicts?

**Comment by FASTSHIFT** [comment]
@kisvegabor 
It seems that the codes generated by `lv_conf_internal_gen.py` and `code-format.py` are not consistent

**Comment by kisvegabor** [comment]
@xiaoxiang781216 made some updates on this but it seems it still not perfect.

**Comment by kisvegabor** [comment]
@mariotaku @tanaytekin

If there were an option to automatically snapshot all widgets we could hugely simplify the SDL and OpenGLES engines. We could let LVGL to SW render everything into buffers and when it draws the pre-rendered images the GPUs can save them into texture and use those textures later on.

To make it work effectively maybe it'd be better to the snapshot the widgets without their children. This way when a children is updated the parent doesn't needs to be redrawn. 

@FASTSHIFT 
I took a second look at `lv_obj_draw_cache_invalidate()`. I think it should work the other way around. `lv_obj_invalidate` should call it automatically, to mark the widget for re-snapshoting.  What do you think?

**Comment by FASTSHIFT** [comment]
> @mariotaku @tanaytekin
> 
> If there were an option to automatically snapshot all widgets we could hugely simplify the SDL and OpenGLES engines. We could let LVGL to SW render everything into buffers and when it draws the pre-rendered images the GPUs can save them into texture and use those textures later on.
> 
> To make it work effectively maybe it'd be better to the snapshot the widgets without their children. This way when a children is updated the parent doesn't needs to be redrawn.
> 
> @FASTSHIFT I took a second look at `lv_obj_draw_cache_invalidate()`. I think it should work the other way around. `lv_obj_invalidate` should call it automatically, to mark the widget for re-snapshoting. What do you think?

For moving position and rotation and scaling, there is no need to redraw. We may need to expand the parameters of `lv_obj_invalidate()` to reduce unnecessary redraw operations.

File: src/core/lv_obj_draw_cache.c (hunk 1 above)
Comment (FASTSHIFT) [reply src/core/lv_obj_draw_cache.c]: Done

File: src/core/lv_obj_draw_cache.c (hunk 2 above)
Comment (FASTSHIFT) [reply src/core/lv_obj_draw_cache.c]: Done

**Comment by mariotaku** [comment]
I have similar concern like @FASTSHIFT . Should a label cache its content including glyphs rendered? Should a button cache every animation frame when it gets pressed? The ability to first render to a buffer then draw it will be very useful for a scrollable list, but extra care should be taken when we decide what to cache and what to render directly...

**Comment by FASTSHIFT** [comment]
I invalidate draw cache by default in `lv_obj_invalidate()`.
Added a `lv_obj_draw_cache_invalidate_cancel()` API cancel invalidate in `set_style_x/y`.

File: src/core/lv_obj_style_gen.c [hunk 3]
```diff
@@ -54,6 +55,9 @@ void lv_obj_set_style_x(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selec
         .num = (int32_t)value
     };
     lv_obj_set_local_style_prop(obj, LV_STYLE_X, v, selector);
+#if LV_USE_OBJ_DRAW_CACHE
```
Comment (kisvegabor) [inline src/core/lv_obj_style_gen.c:58]: This file is auto generated so we can't edit it.

File: src/core/lv_obj_pos.c [hunk 4]
```diff
@@ -862,7 +863,9 @@ void lv_obj_invalidate(const lv_obj_t * obj)
     obj_coords.y2 += ext_size;
 
     lv_obj_invalidate_area(obj, &obj_coords);
-
+#if LV_USE_OBJ_DRAW_CACHE
+    lv_obj_draw_cache_invalidate(obj);
```
Comment (kisvegabor) [inline src/core/lv_obj_pos.c:867]: It should be added rather to `lv_obj_invalidate_area()` as this the last invalidate-like function in the chain.

File: src/core/lv_obj_draw_cache.c [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+void lv_obj_draw_cache_invalidate_cancel(lv_obj_t * obj)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return;
+    }
+
+    obj->spec_attr->draw_cache->invalid = false;
```
Comment (kisvegabor) [inline src/core/lv_obj_draw_cache.c:122]: I'm not sure this is the right approach. If an object's `draw_cache` was already correctly invalidated (e.g. style changed) and it's immediately followed by a position change the original invalidation will be cancelled too.

**Review by kisvegabor — COMMENTED** [review]
I think the best would be to add something like this to the two `lv_obj_invalidate`-s in `lv_obj_pos.c` `lv_obj_move_to()`:
```c
bool draw_cache_inv_old = obj->spec_attr->draw_cache->invalid;
/*Invalidate the original area*/
lv_obj_invalidate(obj);

....
/*Invalidate the new area*/
lv_obj_invalidate(obj);

/*Position change shouldn't cause cache invalidation so restore it*/
obj->spec_attr->draw_cache->invalid = draw_cache_inv_old;
```

**Comment by kisvegabor** [comment]
> The ability to first render to a buffer then draw it will be very useful for a scrollable list, but extra care should be taken when we decide what to cache and what to render directly...

It'd require the implement the whole drawing layer for SDL to support cases when caching is not optimal. Or do you mean just skipping same "texture level caching" when a snapshot is likely to be rendered only once?

**Comment by tanaytekin** [comment]
> @mariotaku @tanaytekin
> 
> If there were an option to automatically snapshot all widgets we could hugely simplify the SDL and OpenGLES engines. We could let LVGL to SW render everything into buffers and when it draws the pre-rendered images the GPUs can save them into texture and use those textures later on.
> 
> To make it work effectively maybe it'd be better to the snapshot the widgets without their children. This way when a children is updated the parent doesn't needs to be redrawn.
> 
> @FASTSHIFT I took a second look at `lv_obj_draw_cache_invalidate()`. I think it should work the other way around. `lv_obj_invalidate` should call it automatically, to mark the widget for re-snapshoting. What do you think?

As far as I know, minimizing the cpu-gpu communication is huge plus. Upload textures once and only set transformation via matrices using shaders when you need to draw. I think for that need to implement shadows - filters etc. (This seems like a very difficult job, I don't know if it's possible) in shader side as well.

**Comment by kisvegabor** [comment]
> I think for that need to implement shadows - filters etc. (This seems like a very difficult job, I don't know if it's possible) in shader side as well.

Why? The idea is to make the SW render draw everything into buffer and the GPU gets a ready to use pixel arrays (bitmaps). So no shadow or other magic needs to be implemented in the GPU, only caching the bitmaps into textures and blending them.

**Comment by mariotaku** [comment]
@kisvegabor I would implement drawing with cached object, as well as the way it draws (primitive elements like rectangles, lines) now. So when the object is rapidly changing, we can still have a good performance.

For other rendering backends, they can also choose to implement one or both. If they only implement drawing with cached object, the object is rendered by CPU. When both are implemented, the object will be rendered by GPU on an off-screen texture.

**Comment by kisvegabor** [comment]
Okay, your call. :slightly_smiling_face: Especially as the SDL rendering is quite full featured. 

What about OpenGL ES? Wouldn't it worth to implement this mixed solution to have something and add more GPU rendering features as a separate step?

**Comment by mariotaku** [comment]
@kisvegabor This would be great. I'll think about a way to introduce "texture" type in lvgl for easier GPU render implementation after this PR finished.

**Comment by kisvegabor** [comment]
#3294 will make it simpler to define new image formats.

**Comment by tanaytekin** [comment]
> > I think for that need to implement shadows - filters etc. (This seems like a very difficult job, I don't know if it's possible) in shader side as well.
> 
> Why? The idea is to make the SW render draw everything into buffer and the GPU gets a ready to use pixel arrays (bitmaps). So no shadow or other magic needs to be implemented in the GPU, only caching the bitmaps into textures and blending them.


> > I think for that need to implement shadows - filters etc. (This seems like a very difficult job, I don't know if it's possible) in shader side as well.
> 
> Why? The idea is to make the SW render draw everything into buffer and the GPU gets a ready to use pixel arrays (bitmaps). So no shadow or other magic needs to be implemented in the GPU, only caching the bitmaps into textures and blending them.

If the filters are changed in a animation for example, need to create textures every frame.(Not ideal)  Other than that, the idea seems pretty good. Also it might be helpful to look at [imgui](https://github.com/ocornut/imgui) they may have implemented it in a similar way.

**Comment by FASTSHIFT** [comment]
@kisvegabor 
Can we restart this PR ? :)

**Comment by FASTSHIFT** [comment]
The [cacheable-container](https://support.touchgfx.com/4.18/docs/development/ui-development/scenarios/achieving-better-performance-with-cacheable-container) of TouchGFX is manually updated by the user, can we also consider a similar approach? This makes the design simpler and easier for users to control.

**Comment by kisvegabor** [comment]
I agree to have this feature. But please wait a little with the implementation. There are too many moving parts now and it's already difficult and error prone to marge the feature branches. 

[Parallel rendering](https://github.com/lvgl/lvgl/issues/3700) improves how "layers" work and I think we should build this feature on top of it. E.g. "layer caching" or so.

**Pull request closed without merging**